### ISSUE TITLE
Allow per-language Git repositories

### ISSUE THREAD

**Issue by m-aciek** [WeblateOrg/weblate#7271]
### Describe the problem

I'm considering [switching with Polish translation of Python documentation](https://github.com/python/python-docs-pl/issues/21) to Weblate. Python documentation translations in different languages are being kept in [separate](https://github.com/python/python-docs-zh-cn/tree/3.10) [Git](https://github.com/python/python-docs-fr) [repositories](https://github.com/python/python-docs-pl).

PO files in the repositories are stored in the main directory, without language prefix in directory structure.

"Filemask" field validation in new component form prevents entering a file mask without language code in it, which would be desired in my case. Also after creating a component potentially some other language would like to switch to using Weblate, and then we'd want those translations to land in separate Git repository, not Polish one.

### Describe the solution you'd like

* Allow to by-pass filemask validation for language placeholder in the pattern.
* I think it's ok for me to have a Weblate project per language for case of Python documentation. It would be awesome to have a possibility for having single project with many repositories for different languages. Presumably that can be difficult to implement, as for as far as I know it's not compatible with current Weblate projects architecture.

### Describe alternatives you've considered

I've [forked original Polish Python docs translations repo](https://github.com/m-aciek/python-docs-pl-weblate-poc) and moved all resources to `pl` directory to set proof-of-concept project on Weblate and investigate other parts of integration.
But such a move is not an option to original repository due to the [convention for fetching and rendering the docs](https://www.python.org/dev/peps/pep-0545/#repository-for-po-files).

### Screenshots

<img src="https://user-images.githubusercontent.com/9288014/154410377-bfbd08c9-16be-4f0d-a8d8-677c0fb3dd7e.png" width="400" alt="Screenshot of validation error displayed after entering about.po in the file mask field">

### Additional context

[PEP 545 -- Python Documentation Translations](https://www.python.org/dev/peps/pep-0545/)

**Comment by josix** [comment]
+ 1

**Comment by nijel** [comment]
Isn't it better for all Python documentation localization to be in a single place? This brings numerous benefits - for example, it is easier to find problematic source strings or people understanding more languages can benefit from other translations.

We do not accept single language projects in our libre hosting plan for this reason.

**Comment by m-aciek** [comment]
Yes, definitely it would be better to have one Weblate project with multiple languages. I consider having separate projects for languages as a temporary workaround only.

**Comment by nijel** [comment]
The problem here is that the "temporary workaround" means quite some development in Weblate. I don't mind getting a pull request implementing this, but it is certainly not something we would push on our own.

**Comment by burner1024** [comment]
you can create a dir and symlink the files into it

**Comment by m-aciek** [comment]
@burner1024 I created a symlink to the main directory, but it didn't work. Please see #7909. I think that creating a `pl` directory and many symlinks to resources instead would be very similar, and also wouldn't work.

**Comment by m-aciek** [comment]
Isn't this issue a duplicate of #5447?

**Comment by nijel** [comment]
No, this is about translating files in a single language repository, that one is about supporting configuration for language teams.

**Comment by m-aciek** [comment]
I'd like to try to tackle this issue in my free time. Instead of doing shenanigans to sync commits from Weblate backend repo to many per-language repositories outside of Weblate, I'd like to invest my time by bringing this feature in Weblate.

For now I imagine that as being able to put a placeholder in repository URL instead in filemask when configuring a component. Then repository state for such a component would be tracked on per language basis. I would like to try to do minimal changes in UX/UI as well as in the code logic to make this all work.

**Comment by nijel** [comment]
So we're looking for support of many single-language repositories within a component, right?

I think the cleanest approach from the current Weblate architecture would be to handle this within the VCS backend.

* It would check out multiple repositories in the working directory. This list should be manually configured and not use wildcards.
* The component would then be configured with an additional prefix.
* All operations would be dispatched to the actual backend for each repository.
* It should not be limited to languages only, there might be other use cases such as https://github.com/WeblateOrg/weblate/issues/10164.

Tricky parts:

* Configuration doesn't support customizing VCS configuration per backend. So for the initial implementation, I'd go with encoding all information in the repository URL. Maybe using JSON, so that UI can be built for that independently.
* Will there be a single multi-repo implementation supporting all VCS integrations (again, this would have to be configured somehow), a separate multi-repo VCS for each existing VCS, or multi-repo would be in the base class and turned on based on JSON in the URL?

The Python docs could then have a repository URL like `{"pl": "https://github.com/python/python-docs-pl.git", "fr": "https://github.com/python/python-docs-pl.git"}` and components would be configured like `*/about.po`.

Do I miss something? Does it sound viable?

**Comment by m-aciek** [comment]
Thank you for your guidance on the direction.

> Will there be a single multi-repo implementation supporting all VCS integrations (again, this would have to be configured somehow), a separate multi-repo VCS for each existing VCS, or multi-repo would be in the base class and turned on based on JSON in the URL?

For now I'm in favour of a single generic multi-repo implementation supporting all VCS integrations.

Then repository JSON could look like `{"pl": {"vcs": "git", "repo": "https://github.com/python/python-docs-pl.git"}, "fr": {"vcs": "git", "repo": "https://github.com/python/python-docs-fr.git"}}`.

If I'm not mistaken it looks like the same format should be repeated for *push* repository URL.

Also probably we would need a dedicated key for "new base" (POTs) repository in the repositories JSON, to use it along with msgmerge addon. Not to hack the dispatch for this case and store the POTs in a language repository.

`{"pl": {"vcs": "git", "repo": "https://github.com/python/python-docs-pl.git"}, "fr": {…}, "new_base": {"vcs": "git", "repo": "https://github.com/python/python-docs-base.git"}`.

**Comment by nijel** [comment]
Yes, that's why I didn't want to bind it to languages only. Using more complex JSON makes sense.

**Comment by nijel** [comment]
Revisiting this again, and I think a different approach might also work after recent changes in Weblate. With increased focus on languages instead of components, having a separate component for each language might be a viable approach as well. The problematic bit would be taking new_base from another repository.

**Comment by amCap1712** [comment]
If we do one component per language, then according to my understanding new base shouldn't be needed. It won't be possible to add a new language to the single language component and you would have to create a new component for a new language/repository. Let me know if I am missing something.

**Comment by nijel** [comment]
It is necessary for invoking the msgmerge add-on. The updated strings come from another repository (which contains `.pot` files), and the translation repositories (containing only `.po` files) should be updated by the msgmerge add-on.

Also, the language repositories seem to start without a translation files (see https://github.com/python/python-docs-sv) so the components approach probably wouldn't work that easy as we would need a way to create new files in the repository for that.

@m-aciek Can you please clarify how updating `.po` files is supposed to work? Is my understanding correct?

**Comment by rffontenelle** [comment]
@nijel Python docs use per-branch versions. So main (if existent should hold only scripts, CI workflows, README and other landing page stuff. The translation of the newly-established Swedish team is happening on [3.14]*https://github.com/python/python-docs-sv/tree/3.14) branch.

Using msgmerge add-ons works just fine: the POT could be generated with `make gettext` using CPython's Makefile and then pushed to Weblate, and have the add-on updating the PO files.

**Comment by nijel** [comment]
I'll rephrase my question: are the repositories standalone or they depend on some other? Like pot files being elsewhere and those would be needed by Weblate.

**Comment by rffontenelle** [comment]
Currently, they don't store POT files. The workflow is to generate POT and update the PO files. Is it possible to have POT files elsewhere or in another branch in the same repository?

**Comment by m-aciek** [comment]
@nijel @rffontenelle We would need a separate repository, like python-docs-gettext or python-docs-pot for storing the “new base” files, and I think we are fine with that. (afaic it's possible to have it in a directory of some language repo, but for the sake of better discoverability, I think it'd be best to have them separately)

Also, I think that we would be ready to start using repositories to store all translations, without [the file treshold](https://github.com/WeblateOrg/weblate/issues/3436), so this shouldn't be an issue.

If I understand correctly the only downside of such integration (separate component for language) could be that e.g. [source checks](https://docs.weblate.org/en/latest/user/checks.html#source-checks) may be ignored multiple times on per-language basis and there would be a bit more of configuration.

**Comment by nijel** [comment]
Yes, dealing with source strings is something that would have to be addressed in this case; we would rather not store them separately for all the languages. This would make some source string checks ineffective, waste the storage, and for billable projects, this would also increase cost because the source language strings would be counted for each component, not just once. I'm just exploring possibilities because the multi-repository setup has its challenges as well.

**Comment by nijel** [comment]
### Multi-repository approach

- All additional logic would be at the VCS integration level.
- The VCS configuration will map multiple repositories into a single tree that will  then be used by Weblate.
- It will be a bit similar to git submodules, just without tracking all repos in a single one.
- It is mostly self-contained and not touching other code.
- The component discovery add-on will work out of the box on this.
- Achieving more complex setups with differing workflows per language might be challenging.

### Multi-component approach

- The component setting would include *language* and *source component* settings. It will be possible to avoid `*` in the file mask then.
- The logic of several blocks would have to be adjusted (source strings handling, including cleanup, source checks trigger, component discovery, ...).
- Each language can have a different configuration, providing great flexibility, but that might make it challenging to maintain (keep all the components in sync).

### Summary

I think sticking with the multi-repository approach is better for this use case, as per-language components will be more challenging to maintain.
